saleor/product/utils/costs.py

```diff
@@ -21,51 +21,59 @@ def __post_init__(self):
 
 
 def get_product_costs_data(
-    product: "Product",
+    product_channel_listing: "ProductChannelListing", channel_slug,
 ) -> Tuple[MoneyRange, Tuple[float, float]]:
 
     purchase_costs_range = MoneyRange(start=zero_money(), stop=zero_money())
     margin = (0.0, 0.0)
 
+    product = product_channel_listing.product
     if not product.variants.exists():
         return purchase_costs_range, margin
 
-    variants = product.variants.all()
-    costs_data = get_cost_data_from_variants(variants)
+    variants = product.variants.all().values_list("id", flat=True)
+    channel_listing = ProductVariantChannelListing.objects.filter(
```

Comment:
`get_product_costs_data` is used in a field resolver in `ProductChannelListing` type. If I fetch N channel listings for a product and for each ask for margin and cost price, I'll have 2xN database queries here.

Reply:
In this case we need to add dataloader. We delay adding dataloaders for channel listing for now. Dataloaders will be added later. I added comment in the code